codalab/apps/teams/forms.py

```diff
@@ -89,3 +95,144 @@ def clean(self):
+        self.instance.status = TeamStatus.objects.get(codename=TeamStatus.APPROVED)
+
+        # We need to call save before adding members
+        self.instance.save()
+
+        if hasattr(self, 'user_list'):
+            # On save, if we have existing objects, clear it. We only add names in the form.
+            if len(self.instance.members.all()) > 0:
+                # Remove users no longer in the list
+                for user in self.instance.members.all():
+                    if user not in self.user_list:
+                        # Delete members not on the new list
+                        print("Deleting membership for user: {}".format(user))
+                        TeamMembership.objects.get(user=user, team=self.instance).delete()
+            # Then we loop through the names/emails they gave us.
+            for user in self.user_list:
+                if user not in self.instance.members.all():
+                    # There is not already an existing membership for this user.
+                    new_team_membership = TeamMembership.objects.create(
+                        user=user,
+                        team=self.instance,
+                        is_invitation=False,
+                        is_request=False,
+                        start_date=timezone.now(),
+                        end_date=timezone.now() + datetime.timedelta(days=9000),
+                        message="Organizer Created",
+                        status=TeamMembershipStatus.objects.get(codename=TeamMembershipStatus.APPROVED),
+                        reason="Organizer Created Team"
+                    )
+                    print("Created new membership: {0} for user: {1} on team: {2}".format(new_team_membership, user, self.instance))
+                else:
+                    print("Existing membership found for user on team.")
+        return super(OrganizerTeamForm, self).save(commit=commit)
+
+
+class OrganizerTeamsCSVForm(forms.Form):
+
+    csv_file = forms.FileField(required=False)
+
+    def __init__(self, *args, **kwargs):
+        self.competition = Competition.objects.get(pk=kwargs.pop('competition_pk'))
+        self.creator = ClUser.objects.get(pk=kwargs.pop('creator_pk'))
+        self.teams_dict = {}
+        super(OrganizerTeamsCSVForm, self).__init__(*args, **kwargs)
+
+    def clean_csv_file(self):
+        teams_dict = {}
+        if self.cleaned_data.get('csv_file'):
+            temp_csv = self.cleaned_data.get('csv_file')
+            csv_team_list = [line.rstrip() for line in temp_csv]
+            for index, value in enumerate(csv_team_list):
+                temp_line = value.split(',')
+                temp_line = [s.strip() for s in temp_line]
+                teams_dict[temp_line.pop(0)] = temp_line
+        self.teams_dict = teams_dict
+
+        for team_name in self.teams_dict:
+            if Team.objects.filter(name=team_name):
+                raise ValidationError("A team with that name already exists!")
```

Comment:
This does a shitload of queries potentially... and it works across all competitions?

Also, we should allow teams that already exist. You should be able to re-run the same CSV over and over without hurting anything?

Reply:
We cannot do that unless we change the team name, cancel the operation, or give the user a choice due to names being unique together with competition on teams. And yeah that query should be narrower, on the same competition.

```unique_together = (('name', 'competition'),)```

Reply:
Detect that the team exists and augment the members to match what is given?

Reply:
You can do `update_or_create` ?